File: plugin_common/src/main/java/com/sn/plugin_common/utils/log/LogUtil.java
package com.sn.plugin_common.utils.log;


import android.text.TextUtils;


import com.sn.plugin_common.BuildConfig;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;


/**
 */
public class LogUtil {

    public static final String TAG = "AccountBook.xlog";

    static {
        System.loadLibrary("c++_shared");
        System.loadLibrary("marsxlog");
    }

    public static void initLog(String logPath,String cachePath) {
//        boolean debug = isLogDebug();
        Xlog.appenderOpen(Xlog.LEVEL_DEBUG, Xlog.AppednerModeAsync, cachePath, logPath, TAG, "");
//        if (BuildConfig.DEBUG) {
//            Xlog.appenderOpen(Xlog.LEVEL_DEBUG, Xlog.AppednerModeAsync, cachePath, logPath, TAG, "");
//            Xlog.setConsoleLogOpen(true);
//
//        } else {
//            Xlog.appenderOpen(Xlog.LEVEL_INFO, Xlog.AppednerModeAsync, cachePath, logPath, TAG, "");
//            Xlog.setConsoleLogOpen(false);
//        }
//
//        Log.setLogImp(new Xlog());
//        Log.appenderFlush(true);
//        android.util.Log.w(TAG, String.format("initLog debug: %b, logPath: %s, cachePath: %s, fileName: %s", false, logPath, cachePath, TAG));
    }

    /**
     * 判断本地是否开启调试日志
     *
     * @return true开启
     */
//    private static boolean isLogDebug() {
//        BufferedReader reader = null;
//        try {
//            File conf = new File(FileAccessor.APPS_ROOT_DIR + File.separator + "logConfig.cnf");
//            if (!conf.exists()) {
//                boolean ret = conf.createNewFile();
//                android.util.Log.i(TAG, "createNewFile ret:" + ret);
//            }
//            reader = new BufferedReader(new FileReader(conf));
//            String line = reader.readLine();
//            if (!TextUtil.isEmpty(line)) {
//                return "true".equals(line);
//            }
//        } catch (Exception e) {
//            e.printStackTrace();
//        } finally {
//            CloseQuality.close(reader);
//        }
//        return false;
//    }

    /**
     * 获取进程号对应的进程名
     *
     * @return 进程名
     */
    public static String getProcessName() {
        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new FileReader("/proc/" + android.os.Process.myPid() + "/cmdline"));
            String processName = reader.readLine();
            if (!TextUtils.isEmpty(processName)) {
                processName = processName.trim();
            }
            String[] split = processName.split(":");
            if (split.length == 2) {
                return "[" + split[1] + "]";
            }
            return "";
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        } finally {
            try {
                if (reader != null) {
                    reader.close();
                }
            } catch (IOException exception) {
                exception.printStackTrace();
            }
        }
        return "";
    }

    public static void close() {
        Log.appenderClose();
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    public static void v(String msg) {
        v(TAG, msg);
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    public static void v(String tag, String msg) {
        v(tag, msg, (Object[]) null);
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    public static void v(String tag, final String format, final Object... obj) {
        Log.v(tag, format, obj);
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    public static void d(String msg) {
        d(TAG, msg);
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    public static void d(String tag, String msg) {
        d(tag, msg, (Object[]) null);
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    public static void d(String tag, final String format, final Object... obj) {
        Log.d(tag, format, obj);
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    public static void i(String msg) {
        i(TAG, msg);
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    public static void i(String tag, String msg) {
        i(tag, msg, (Object[]) null);
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    public static void i(String tag, final String format, final Object... obj) {
        Log.i(tag, format, obj);
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    public static void w(String msg) {
        w(TAG, msg);
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    public static void w(String tag, String msg) {
        w(tag, msg, (Object[]) null);
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    public static void w(String tag, final String format, final Object... obj) {
        Log.w(tag, format, obj);
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    public static void e(String msg) {
        e(TAG, msg);
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    public static void e(String tag, String msg) {
        e(tag, msg, (Object[]) null);
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    public static void e(String tag, final String format, final Object... obj) {
        Log.e(tag, format, obj);
    }

    /**
     * 打印程序异常日志
     *
     * @param tag    日志tag
     * @param tr     异常信息
     * @param format 格式化串
     * @param obj    格式化占位
     */
    public static void printErrStackTrace(String tag, Throwable tr, final String format, final Object... obj) {
        Log.printErrStackTrace(tag, tr, format, obj);
    }


    /**
     * 打印程序堆栈日志
     *
     * @param tag 日志tag
     */
    public static void printStackTrace(String tag) {
        StackTraceElement[] element = Thread.currentThread().getStackTrace();
        StringBuilder sb = new StringBuilder(512);
        sb.append("************************************************************").append("\n");
        for (StackTraceElement traceElement : element)
            sb.append("\tat " + traceElement).append("\n");
        ;
        sb.append("************************************************************");
        android.util.Log.e(tag, sb.toString());
    }


    public static String getLogUtilsTag(Class<? extends Object> clazz) {
        return LogUtil.TAG + "." + clazz.getSimpleName();
    }
}

File: plugin_common/src/main/java/com/gx/utils/log/LogUtil.kt
package com.gx.utils.log

import android.content.Context
import android.os.Build.MODEL
import android.os.Build.VERSION.SDK_INT
import android.os.Process
import android.text.TextUtils
import com.gx.plugin_common.BuildConfig
import com.gx.utils.apk.AppInfoUtil.getAppName
import com.gx.utils.apk.AppInfoUtil.getVersionCode
import com.gx.utils.apk.AppInfoUtil.getVersionName
import com.tencent.mars.xlog.Log
import com.tencent.mars.xlog.Xlog
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

/**
 *
 */
object LogUtil {
    const val TAG = "LogUtil"
    fun initLog(logPath: String?, cachePath: String?) {
        val debug = isLogDebug
        if (BuildConfig.DEBUG || debug) {
            Xlog.appenderOpen(
                Xlog.LEVEL_DEBUG,
                Xlog.AppednerModeAsync,
                cachePath,
                logPath,
                TAG,
                ""
            )
            Xlog.setConsoleLogOpen(true)
        } else {
            Xlog.appenderOpen(
                Xlog.LEVEL_INFO,
                Xlog.AppednerModeAsync,
                cachePath,
                logPath,
                TAG,
                ""
            )
            Xlog.setConsoleLogOpen(false)
        }
        Log.setLogImp(Xlog())
        Log.appenderFlush(true)
        android.util.Log.w(
            TAG,
            String.format(
                "initLog debug: %b, logPath: %s, cachePath: %s, fileName: %s",
                false,
                logPath,
                cachePath,
                TAG
            )
        )
    }

    /**
     * 判断本地是否开启调试日志
     * 通过本地配置文件开启日志
     *
     * @return true开启
     */
    private val isLogDebug: Boolean
        private get() = true

    /**
     * 获取进程号对应的进程名
     *
     * @return 进程名
     */
    val processName: String
        get() {
            var reader: BufferedReader? = null
            try {
                reader =
                    BufferedReader(FileReader("/proc/" + Process.myPid() + "/cmdline"))
                var processName = reader.readLine()
                if (!TextUtils.isEmpty(processName)) {
                    processName = processName.trim { it <= ' ' }
                }
                val split = processName.split(":".toRegex()).toTypedArray()
                return if (split.size == 2) {
                    "[" + split[1] + "]"
                } else ""
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            } finally {
                try {
                    reader?.close()
                } catch (exception: IOException) {
                    exception.printStackTrace()
                }
            }
            return ""
        }

    fun close() {
        Log.appenderClose()
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    fun v(msg: String?) {
        v(TAG, msg)
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    fun v(tag: String?, msg: String?) {
        v(tag, msg, null)
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    fun v(tag: String?, format: String?, vararg obj: Any?) {
        Log.v(tag, format, *obj)
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    fun d(msg: String?) {
        d(TAG, msg)
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    fun d(tag: String?, msg: String?) {
        d(tag, msg, null)
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    fun d(tag: String?, format: String?, vararg obj: Any?) {
        Log.d(tag, format, *obj)
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    fun i(msg: String?) {
        i(TAG, msg)
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    fun i(tag: String?, msg: String?) {
        i(tag, msg, null)
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    fun i(tag: String?, format: String?, vararg obj: Any?) {
        Log.i(tag, format, *obj)
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    fun w(msg: String?) {
        w(TAG, msg)
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    fun w(tag: String?, msg: String?) {
        w(tag, msg, null)
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    fun w(tag: String?, format: String?, vararg obj: Any?) {
        Log.w(tag, format, *obj)
    }

    /**
     * 使用默认的TAG 打印日志
     *
     * @param msg 日志内容
     */
    fun e(msg: String?) {
        e(TAG, msg)
    }

    /**
     * 根据提高的日志TAG和msg内容打印日志
     *
     * @param tag 日志tag
     * @param msg 日志内容
     */
    fun e(tag: String?, msg: String?) {
        e(tag, msg, null)
    }

    /**
     * 根据提供的日志TAG进行日志格式化输出
     *
     * @param tag    日志tag
     * @param format 格式化占位符
     * @param obj    格式化占位符替代内容
     */
    fun e(tag: String?, format: String?, vararg obj: Any?) {
        Log.e(tag, format, *obj)
    }

    /**
     * 打印程序异常日志
     *
     * @param tag    日志tag
     * @param tr     异常信息
     * @param format 格式化串
     * @param obj    格式化占位
     */
    fun printErrStackTrace(
        tag: String?,
        tr: Throwable?,
        format: String?,
        vararg obj: Any?
    ) {
        Log.printErrStackTrace(tag, tr, format, *obj)
    }

    /**
     * 打印程序堆栈日志
     *
     * @param tag 日志tag
     */
    fun printStackTrace(tag: String?) {
        val element =
            Thread.currentThread().stackTrace
        val sb = StringBuilder(512)
        sb.append("************************************************************").append("\n")
        for (traceElement in element) sb.append("\tat $traceElement")
            .append("\n")
        sb.append("************************************************************")
        android.util.Log.e(tag, sb.toString())
    }

    fun getLogUtilsTag(clazz: Class<out Any?>): String {
        return TAG + "." + clazz.simpleName
    }

    /**
     * 打印设备&APP 信息
     *
     * @param context
     */
    fun printHeader(context: Context?) {
        val versionName = getVersionName(context!!)
        val appName = getAppName(context)
        val versionCode = getVersionCode(context)
        val sb = StringBuilder()
        sb.append("\r\n")
        sb.append(" ---------- $appName  INFO  ---------------")
        sb.append("\r\n")
        sb.append(" DEVICE MODEL:[ $MODEL ] ")
        sb.append("\r\n")
        sb.append(" ANDROID VERSION:[ $SDK_INT ] ")
        sb.append("\r\n")
        sb.append(" APP VERSION NAME:[ $versionName ] ")
        sb.append("\r\n")
        sb.append(" APP VERSION CODE:[ $versionCode ] ")
        sb.append("\r\n")
        i(TAG, sb.toString())
    }

    init {
        System.loadLibrary("stlport_shared")
        System.loadLibrary("marsxlog")
    }
}